src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,170 @@
+                tempArray[index] = list.get(index);
+            }
+            return tempArray;
+        }
+        return null;
+    }
+
+
     @Override
     public void add(T value) {
-
+        int index = cursor;
+        ArrayList.this.add(value, index);
+        cursor = index + 1;
     }
 
+
     @Override
     public void add(T value, int index) {
-
+        if (index > size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException(errorMsg(index));
+        } else {
+            int s = size;
+            if (s == array.length) {
+                array = (T[]) grow();
+            }
+            System.arraycopy(array, index,
+                array, index + 1,
+                s - index);
+            array[index] = value;
+            size = s + 1;
+        }
     }
 
+
     @Override
     public void addAll(List<T> list) {
-
+        if (!list.isEmpty()) {
+            T[] tempArray = (T[]) listToArray(list);
+            int newCapacity = size + tempArray.length;
+            array = Arrays.copyOf(array, newCapacity);
+            System.arraycopy(tempArray, 0, array, size, tempArray.length);
+            size = newCapacity;
+        }
     }
 
+
     @Override
     public T get(int index) {
-        return null;
+        if (index >= size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException(errorMsg(index));
+        } else {
+            return array[index];
+        }
     }
 
+
     @Override
     public void set(T value, int index) {
-
+        if (index >= size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException(errorMsg(index));
+        } else {
+            array[index] = value;
+        }
     }
 
+
     @Override
     public T remove(int index) {
-        return null;
+        if (index > size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException(errorMsg(index));
+        } else {
+            int newSize = size - 1;
+            final T oldValue = get(index);
+            if (newSize > index) {
+                System.arraycopy(array, index + 1, array, index, newSize - index);
+            }
+            size = newSize;
+            array[newSize] = null;
+            return oldValue;
+        }
     }
 
+
     @Override
     public T remove(T element) {
-        return null;
+        int index = 0;
+        found:
+        {
+            if (element == null) {
+                for (; index < size; index++) {
+                    if (array[index] == null) {
+                        remove(index);
+                        break found;
+                    }
+                }
+            } else {
+                for (; index < size; index++) {
+                    if (element.equals(array[index])) {
+                        remove(index);
+                        break found;
+                    }
+                }
+                throw new NoSuchElementException();
+            }
+        }
+        return element;
```

Comment:
We can remove outer if-else block at all and make only one for loop with one if statement. It would be much easier to read and understand. For this purpose we can use Objects util class to simplify statement. Also we can remove return statement at the end and place it into if-statements

```suggestion
        for (int index = 0; index < size; index++) {
            if (Objects.equals(array[index], element)) {
                remove(index);
                return element;
            }
        }

        throw new NoSuchElementException();
```